src/manager.S

```diff
@@ -129,3 +129,67 @@ add_compartment:
     b.gt      abort
 
     ret
+
+/**
+ * Function to delete an existing compartment data
+ *
+ * @param x0 Function of compartment to be deleted
+ */
+/* TODO there is a potential issue of a compartment saving a capability it was
+ * provided to it for use (i.e., PCC, DDC), such as by storing it in the memory
+ * of another compartment. If the original compartment gets deleted, and a new
+ * compartment is allocated in the same memory region, we have leaked a
+ * capability referencing the new compartment. */
+.type del_compartment, "function"
+del_compartment:
+    // Find capabilities (DDC/PCC) of given compartment
+    ldr       x1, comps_addr
+    mov       w2, #0
+    ldr       w3, comps_cnt
+
+pcc_check_start:
+    ldr       c4, [x1, #COMP_OFFSET_PCC] // get current stored PCC to check
+    gcvalue   x4, c4
+    cmp       x4, x0 // check value of current PCC against given function
+    b.eq      pcc_found
+    add       w2, w2, #1
+    cmp       w2, w3 // compartment bounds check
+    b.eq      pcc_not_found
```

Comment:
This is perfectly fine, but I can offer some assembly tips, in case they're useful.

This only uses `w2` as  loop counter, so it's (slightly) more efficient to count down:

```
    ...  # Don't use w2 at all.
    ldr  w3, comps_cnt

pcc_check_start:
    ...
    subs  w3, w3, #1     # Sets Z flag when the result is zero.
    b.eq  pcc_not_found   # 'eq' tests the Z flag, as if asking "was w3 equal to 1?"
```

That assumes that `comps_cnt` is not zero, as does the original. If that's a problem, you can move the check (and branch) to the start and use `b.mi` (branching on negative resutls) to avoid the off-by-one issue that a verbatim move would introduce.

Reply:
I've tried implementing what you propose in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/2/commits/e106bb036bd28231ff02f7bf4b3726296d422ec3. Do you think this is sufficient?

Reply:
Oh, I hadn't replied here. It had an issue, reported on that revision, but is fine now. I'll resolve this conversation.